classes/security/Validation.inc.php

```diff
@@ -263,15 +307,55 @@ static function generatePassword($length = 8) {
 	/**
 	 * Generate a hash value to use for confirmation to reset a password.
 	 * @param $userId int
+	 * @param $expiry int timestamp when hash expires, defaults to CURRENT_TIME + RESET_SECONDS
 	 * @return string (boolean false if user is invalid)
 	 */
-	static function generatePasswordResetHash($userId) {
+	static function generatePasswordResetHash($userId, $expiry = null) {
 		$userDao = DAORegistry::getDAO('UserDAO');
 		if (($user = $userDao->getById($userId)) == null) {
 			// No such user
 			return false;
 		}
-		return substr(md5($user->getId() . $user->getUsername() . $user->getPassword()), 0, 6);
+		// create hash payload
+		$salt = Config::getVar('security', 'salt');
+
+		if (empty($expiry)) {
+			$expires = (int) Config::getVar('security', 'reset_seconds', 7200);
+			$expiry = time() + $expires;
+		}
+
+		// use last login time to ensure the hash changes when they log in
+		$data = $user->getUsername() . $user->getPassword() . $user->getDateLastLogin() . $expiry;
+
+		// generate hash and append expiry timestamp
+		$algos = hash_algos();
+
+		foreach (array('sha256', 'sha1', 'md5') as $algo) {
```

Comment:
I'm a little foggy on the details, but this is MD5 using HMAC and the PHP `hmac` function set, vs. plain old MD5 using a legacy function. They will behave a little differently but it's harmless, I think.